perf: Optimize RDF queries

bprusinowski opened the pull request.

Rdataflow commented:
@bprusinowski FYI incorrect datasetcount you observed might be related to missing `dcterms:creator` property or the like...

action points
- enforce mandatory properties in validation profile (probably will be part in opendata.swiss profile) -> validation
- ensure consistency of multiple queries on your side (i.e. same properties mandatory) -> ITX

bprusinowski commented:
Hi @Rdataflow, I think the PR is ready for review.

I didn't remove themes and organizations queries, as we rely on them to compute filters to query the cubes. Maybe they could be removed, but they serve as a validation, to make sure our filters are correct. Now that the performance of cubes query, and remaining two for themes and organizations, that might not be a big problem, as they are fired in parallel anyway.

Let me know what you think!

It looks that the performance of fetching the cubes improved significantly (I didn't test in a sophisticated way, just by refreshing the PR and PROD pages several times with the same set of filters).

PROD
**No filters**: ~2s
**Theme & organizations filters applied**: ~1.5s
**Theme & organizations filters applied & text search**: ~1.5s

PR
**No filters**: ~0.8s
**Theme & organizations filters applied**: ~0.7s
**Theme & organizations filters applied & text search**: ~0.7s

Rdataflow commented:
Hi @bprusinowski 

Thank you for all your work and this PR. I just run into a first bug:
- open https://visualization-tool-git-perf-optimize-rdf-queries-ixt1.vercel.app/browse?includeDrafts=false&order=SCORE&search=wasser+wald&dataSource=Prod
- this results in an error `[GraphQL] Cannot read properties of undefined (reading 'score')`

One test I propose: (but currently unable to test)
equal results for
- https://visualization-tool-git-perf-optimize-rdf-queries-ixt1.vercel.app/browse?includeDrafts=false&order=SCORE&search=wasser+wald&dataSource=Prod
- https://visualization-tool-git-perf-optimize-rdf-queries-ixt1.vercel.app/browse?includeDrafts=false&order=SCORE&search=wald+wasser&dataSource=Prod

One suggestion:
refactor URl params: new default to ?org=https://...&theme=https://...

Some SPARQL specific details:
- Include `GRAPH <>` directive whereever possible to restrict the query to 1 graph only and subsequently to 1 class (a.k.a. `rdf:type`) and `schema:inDefinedTermSet`
  - [x] creator: `GRAPH <https://lindas.admin.ch/sfa/opendataswiss> { ?creatorIri a schema:Organization ;             schema:inDefinedTermSet <https://register.ld.admin.ch/opendataswiss/org>` ...
  - [ ] theme: `GRAPH <https://lindas.admin.ch/sfa/opendataswiss> { ?themeIri a schema:DefinedTerm ;            schema:inDefinedTermSet <https://register.ld.admin.ch/opendataswiss/category>` ...
  - subtheme: depending on the selected creator (currently only BAFU has subthemes defined)
      - [ ] BAFU: `GRAPH <https://lindas.admin.ch/foen/themes> { ?subthemeIri a schema:DefinedTerm ;            schema:inDefinedTermSet <https://register.ld.admin.ch/foen/theme>` ...
        - FYI: you may also declare `VALUES ( ?subthemeGraph ?subthemeTermset ) { ( <https://lindas.admin.ch/foen/themes> <https://register.ld.admin.ch/foen/theme> ) }` and reuse those values with by their variable name. if in the future more organizations declare subthemes, visualize only needs to enhance the list of possible VALUES `{ ( foo bar ) ... }`
        - ... and the condition `?subthemeIri schema:inDefinedTermSet <https://register.ld.admin.ch/foen/theme> .` therefore should be mandatory - not optional

outlook (FYI):
- `schema:datePublished` datatype is a mess in the data currently. for the future validation will help improving the situation and later visualize can step up, but that's a future PR 😄
- hopefully with the upcoming validation work for opendata.swiss there will some more properties become mandatory, so we could then remove some `OPTIONAL`s ...

that's my first impressions. I can have a deeper look at it by next week... query speed goes into the right direction 👍

bprusinowski commented:
Thanks for taking a look, testing and suggestions @Rdataflow 🙇‍♂️ I'll go through them today 👌

Rdataflow commented:
@bprusinowski one more important thing to keep in mind:
- [ ] ensure to keep multilang filtering in place:
  - test following queries all deliver equal results: (1 dataset as of now)
    - /it/browse?includeDrafts=false&order=SCORE&search=wasser&dataSource=Prod
    - /fr/browse?includeDrafts=false&order=SCORE&search=wasser&dataSource=Prod
    - /browse?includeDrafts=false&order=SCORE&search=wasser&dataSource=Prod
    - /it/browse?includeDrafts=false&order=SCORE&search=water&dataSource=Prod
    - /fr/browse?includeDrafts=false&order=SCORE&search=acque&dataSource=Prod
    - /browse?includeDrafts=false&order=SCORE&search=eaux&dataSource=Prod

at high level I see two possibilites:
- push filter logic to sparql (see example)
- retrieve multilang from sparql (includes 4 times the data, to consider: in the future there will be more data online)

Rdataflow commented:
@bprusinowski GraphQL issue around score is resolved now :+1: would it make sense to include a test for equal results in CI tests? (i.e. wald+wasser=wasser+wald) ... if you didn't include this already (I'm not aware)

bprusinowski commented:
@Rdataflow I think I implemented all the mentioned changes (and more tests) :)

Let me know if you're fine with merging to TEST, so it can be tested there afterwards (on the same / similar infrastructure as PROD) 👍

Rdataflow commented:
@bprusinowski general direction looks good, just some remarks:
- [ ] drop the comment on dates entirely. NB: the date hotfix would be the other way round, but it will be easier anyway to fix that in the data ... 
- [ ] rework keyword filter for better performance: performance is generally better by only joining one triple once (i.e. ?creatorLabel etc.) and then reusing or filtering those variables. in addition no cartesians happen this way. and `VALUES ?keyword {}` can simplify the keyword filtering part...
  - example query:
  ```
  VALUES ?keyword { "baum" "water" "klima" }
  
  FILTER(
    CONTAINS(LCASE(?title_de), LCASE(?keyword)) ||
    CONTAINS(LCASE(?title_fr), LCASE(?keyword)) ||
    CONTAINS(LCASE(?title_it), LCASE(?keyword)) ||
    CONTAINS(LCASE(?title_en), LCASE(?keyword)) ||
    CONTAINS(LCASE(?title_),   LCASE(?keyword)) ||

    CONTAINS(LCASE(?description_de), LCASE(?keyword)) ||
    CONTAINS(LCASE(?description_fr), LCASE(?keyword)) ||
    CONTAINS(LCASE(?description_it), LCASE(?keyword)) ||
    CONTAINS(LCASE(?description_en), LCASE(?keyword)) ||
    CONTAINS(LCASE(?description_),   LCASE(?keyword)) ||
      
    CONTAINS(LCASE(?creatorLabel_de), LCASE(?keyword)) || 
    CONTAINS(LCASE(?creatorLabel_fr), LCASE(?keyword)) || 
    CONTAINS(LCASE(?creatorLabel_it), LCASE(?keyword)) || 
    CONTAINS(LCASE(?creatorLabel_en), LCASE(?keyword)) || 
    CONTAINS(LCASE(?creatorLabel_),   LCASE(?keyword)) || 
      
    CONTAINS(LCASE(?themeLabel_de), LCASE(?keyword)) || 
    CONTAINS(LCASE(?themeLabel_fr), LCASE(?keyword)) || 
    CONTAINS(LCASE(?themeLabel_it), LCASE(?keyword)) || 
    CONTAINS(LCASE(?themeLabel_en), LCASE(?keyword)) || 
    CONTAINS(LCASE(?themeLabel_),   LCASE(?keyword)) || 
      
    CONTAINS(LCASE(?subthemeLabel_de), LCASE(?keyword)) || 
    CONTAINS(LCASE(?subthemeLabel_fr), LCASE(?keyword)) || 
    CONTAINS(LCASE(?subthemeLabel_it), LCASE(?keyword)) || 
    CONTAINS(LCASE(?subthemeLabel_en), LCASE(?keyword)) || 
    CONTAINS(LCASE(?subthemeLabel_),   LCASE(?keyword)) || 
      
    CONTAINS(LCASE(?publisher), LCASE(?keyword))
  )
  ```

t0b3 commented:
@bprusinowski one more thing:
- [ ] add the subthemes relevant parts only if org matches 
  - currently only for `org=https://register.ld.admin.ch/opendataswiss/org/bundesamt-fur-umwelt-bafu` and then inject the matching `VALUES ( ?subthemeGraph ?subthemeTermset )` ... this will further speed up the response time

bprusinowski commented:
Thanks @Rdataflow and @t0b3 for your suggestions, I implemented them in https://github.com/visualize-admin/visualization-tool/pull/1236/commits/7eea7eebabc93081468cbd3777654bb52fcdbdd1.

@Rdataflow I am not sure if I fully understood your comment on keywords, when using the proposed query with

`CONTAINS(LCASE(?title_de), LCASE(?keyword)) ||
  CONTAINS(LCASE(?title_fr), LCASE(?keyword)) ||
  CONTAINS(LCASE(?title_it), LCASE(?keyword)) ||
  CONTAINS(LCASE(?title_en), LCASE(?keyword)) ||
  CONTAINS(LCASE(?title_),   LCASE(?keyword)) || ...`

it returns no results. I updated the query to use `VALUES ( ... ?keyword )` pattern, let me know if it looks ok now, or if there is something else to optimize 👍

Rdataflow commented:
searching for keyword `foo` + `bar` returns `[GraphQL] Bad Request (400): {"message":"com.complexible.stardog.plan.eval.ExecutionException: Invalid SPARQL query: Encountered \" \",\" \", \"\" at line 190, column 32.\nWas expecting one of:\n    \"}\" ...\n    \"<<\" ...\n    \"^^\" ...\n    \"true\" ...\n    \"false\" ...\n    \"UNDEF\" ...\n    <Q_IRI_REF> ...\n    <PNAME_NS> ...\n    <PNAME_LN> ...\n    <LANGTAG> ...\n    <INTEGER> ...\n    <INTEGER_POSITIVE> ...\n    <INTEGER_NEGATIVE> ...\n    <DECIMAL> ...\n    <DECIMAL_POSITIVE> ...\n    <DECIMAL_NEGATIVE> ...\n    <DOUBLE> ...\n    <DOUBLE_POSITIVE> ...\n    <DOUBLE_NEGATIVE> ...\n    <STRING_LITERAL1> ...\n    <STRING_LITERAL2> ...\n    <STRING_LITERAL_LONG1> ...\n    <STRING_LITERAL_LONG2> ...\n    ","code":"QE0PE2","stacktrace":"com.stardog.stark.query.QueryExecutionFailure: com.complexible.stardog.plan.eval.ExecutionException: Invalid SPARQL query: Encountered \" \",\" \", \"\" at line 190, column 32.\nWas expecting one of:\n    \"}\" ...\n    \"<<\" ...\n    \"^^\" ...\n    \"true\" ...\n    \"false\" ...\n    \"UNDEF\" ...\n    <Q_IRI_REF> ...\n    <PNAME_NS> ...\n    <PNAME_LN> ...\n    <LANGTAG> ...\n    <INTEGER> ...\n    <INTEGER_POSITIVE> ...\n    <INTEGER_NEGATIVE> ...\n    <DECIMAL> ...\n    <DECIMAL_POSITIVE> ...\n    <DECIMAL_NEGATIVE> ...\n    <DOUBLE> ...\n    <DOUBLE_POSITIVE> ...\n    <DOUBLE_NEGATIVE> ...\n    <STRING_LITERAL1> ...\n    <STRING_LITERAL2> ...\n    <STRING_LITERAL_LONG1> ...\n    <STRING_LITERAL_LONG2> ...\n    \n\tat com.stardog.stark.query.QueryExecutionFailure.forLoggedError(QueryExecutionFailure.java:46)\n\tat com.complexible.stardog.query.DefaultQueryFactory$SelectQuery.execute(DefaultQueryFactory.java:273)\n\tat com.complexible.stardog.query.DefaultQueryFactory$SelectQuery.execute(DefaultQueryFactory.java:256)\n\tat com.complexible.stardog.query.DelegatingQuery.execute(DelegatingQuery.java:68)\n\tat com.complexible.stardog.db.DatabaseImpl$TimerPausingReadQuery.execute(DatabaseImpl.java:1766)\n\tat com.complexible.stardog.db.DatabaseImpl$TimerPausingReadQuery.execute(DatabaseImpl.java:1753)\n\tat com.complexible.stardog.query.DelegatingQuery.execute(DelegatingQuery.java:60)\n\tat com.complexible.stardog.StardogKernel$SecuredReadQuery.execute(StardogKernel.java:4852)\n\tat com.complexible.stardog.StardogKernel$SecuredReadQuery.execute(StardogKernel.java:4839)\n\tat com.complexible.stardog.query.DelegatingQuery.execute(DelegatingQuery.java:60)\n\tat com.complexible.stardog.MonitoredQuery$MonitoredReadQuery.execute(MonitoredQuery.java:54)\n\tat com.complexible.stardog.MonitoredQuery$MonitoredReadQuery.execute(MonitoredQuery.java:36)\n\tat com.complexible.stardog.protocols.http.server.ProtocolUtils.executeReadQuery(ProtocolUtils.java:544)\n\tat com.complexible.stardog.protocols.http.server.ProtocolUtils.executeReadQuery(ProtocolUtils.java:517)\n\tat com.complexible.stardog.protocols.http.server.SPARQLProtocol.executeQuery(SPARQLProtocol.java:250)\n\tat com.complexible.stardog.protocols.http.server.SPARQLProtocol.getSparqlQuery(SPARQLProtocol.java:139)\n\tat com.stardog.http.server.undertow.jaxrs.ExtractRoutes.lambda$handleIt$5(ExtractRoutes.java:192)\n\tat org.apache.shiro.subject.support.SubjectRunnable.doRun(SubjectRunnable.java:120)\n\tat org.apache.shiro.subject.support.SubjectRunnable.run(SubjectRunnable.java:108)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\nCaused by: com.complexible.stardog.plan.eval.ExecutionException: Invalid SPARQL query: Encountered \" \",\" \", \"\" at line 190, column 32.\nWas expecting one of:\n    \"}\" ...\n    \"<<\" ...\n    \"^^\" ...\n    \"true\" ...\n    \"false\" ...\n    \"UNDEF\" ...\n    <Q_IRI_REF> ...\n    <PNAME_NS> ...\n    <PNAME_LN> ...\n    <LANGTAG> ...\n    <INTEGER> ...\n    <INTEGER_POSITIVE> ...\n    <INTEGER_NEGATIVE> ...\n    <DECIMAL> ...\n    <DECIMAL_POSITIVE> ...\n    <DECIMAL_NEGATIVE> ...\n    <DOUBLE> ...\n    <DOUBLE_POSITIVE> ...\n    <DOUBLE_NEGATIVE> ...\n    <STRING_LITERAL1> ...\n    <STRING_LITERAL2> ...\n    <STRING_LITERAL_LONG1> ...\n    <STRING_LITERAL_LONG2> ...\n    \n\tat com.complexible.stardog.plan.eval.QueryEngine.handleError(QueryEngine.java:308)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executePlan(QueryEngine.java:487)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executePlan(QueryEngine.java:237)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executePlan(QueryEngine.java:451)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executeSelectPlan(QueryEngine.java:406)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executeSelect(QueryEngine.java:389)\n\tat com.complexible.stardog.query.DefaultQueryFactory$SelectQuery.execute(DefaultQueryFactory.java:270)\n\t... 20 more\nCaused by: com.complexible.stardog.plan.parser.QueryParseException: Invalid SPARQL query: Encountered \" \",\" \", \"\" at line 190, column 32.\nWas expecting one of:\n    \"}\" ...\n    \"<<\" ...\n    \"^^\" ...\n    \"true\" ...\n    \"false\" ...\n    \"UNDEF\" ...\n    <Q_IRI_REF> ...\n    <PNAME_NS> ...\n    <PNAME_LN> ...\n    <LANGTAG> ...\n    <INTEGER> ...\n    <INTEGER_POSITIVE> ...\n    <INTEGER_NEGATIVE> ...\n    <DECIMAL> ...\n    <DECIMAL_POSITIVE> ...\n    <DECIMAL_NEGATIVE> ...\n    <DOUBLE> ...\n    <DOUBLE_POSITIVE> ...\n    <DOUBLE_NEGATIVE> ...\n    <STRING_LITERAL1> ...\n    <STRING_LITERAL2> ...\n    <STRING_LITERAL_LONG1> ...\n    <STRING_LITERAL_LONG2> ...\n    \n\tat com.complexible.stardog.plan.parser.QueryParserImpl.parseQuery(QueryParserImpl.java:597)\n\tat com.complexible.stardog.plan.parser.QueryParserImpl.parseQuery(QueryParserImpl.java:553)\n\tat com.complexible.stardog.plan.PlanContext.planFor(PlanContext.java:137)\n\tat com.complexible.stardog.plan.eval.QueryEngine.planFor(QueryEngine.java:598)\n\tat com.complexible.stardog.plan.eval.QueryEngine.lambda$executeSelect$16(QueryEngine.java:389)\n\tat com.complexible.stardog.plan.eval.QueryEngine.getExecutablePlan(QueryEngine.java:722)\n\tat com.complexible.stardog.plan.eval.QueryEngine.executePlan(QueryEngine.java:477)\n\t... 25 more\nCaused by: com.complexible.common.rdf.query.parser.sparql.ast.ParseException: Encountered \" \",\" \", \"\" at line 190, column 32.\nWas expecting one of:\n    \"}\" ...\n    \"<<\" ...\n    \"^^\" ...\n    \"true\" ...\n    \"false\" ...\n    \"UNDEF\" ...\n    <Q_IRI_REF> ...\n    <PNAME_NS> ...\n    <PNAME_LN> ...\n    <LANGTAG> ...\n    <INTEGER> ...\n    <INTEGER_POSITIVE> ...\n    <INTEGER_NEGATIVE> ...\n    <DECIMAL> ...\n    <DECIMAL_POSITIVE> ...\n    <DECIMAL_NEGATIVE> ...\n    <DOUBLE> ...\n    <DOUBLE_POSITIVE> ...\n    <DOUBLE_NEGATIVE> ...\n    <STRING_LITERAL1> ...\n    <STRING_LITERAL2> ...\n    <STRING_LITERAL_LONG1> ...\n    <STRING_LITERAL_LONG2> ...\n    \n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.generateParseException(SyntaxTreeBuilder.java:11256)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.jj_consume_token(SyntaxTreeBuilder.java:11109)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.InlineDataOneVar(SyntaxTreeBuilder.java:1929)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.DataBlock(SyntaxTreeBuilder.java:1879)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.InlineData(SyntaxTreeBuilder.java:1853)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.ScopeTerminator(SyntaxTreeBuilder.java:3266)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2853)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2995)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.GroupGraphPattern(SyntaxTreeBuilder.java:2805)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.GroupOrUnionGraphPattern(SyntaxTreeBuilder.java:3351)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.ScopeTerminator(SyntaxTreeBuilder.java:3250)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2956)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2995)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2910)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Scopes(SyntaxTreeBuilder.java:2995)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.GroupGraphPattern(SyntaxTreeBuilder.java:2805)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.WhereClause(SyntaxTreeBuilder.java:1463)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.SelectQuery(SyntaxTreeBuilder.java:780)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.Query(SyntaxTreeBuilder.java:445)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.QueryContainer(SyntaxTreeBuilder.java:217)\n\tat com.complexible.common.rdf.query.parser.sparql.ast.SyntaxTreeBuilder.parseQuery(SyntaxTreeBuilder.java:31)\n\tat com.complexible.stardog.plan.parser.QueryParserImpl.parseQuery(QueryParserImpl.java:558)\n\t... 31 more\n","stacktraceBinary":"rO0ABXNyAC1jb20uc3RhcmRvZy5zdGFyay5xdWVyeS5RdWVyeUV4ZWN1dGlvbkZhaWx1cmXH1b5kJNTq6AIAAVoAB21Mb2dnZWR4cgAaamF2YS5sYW5nLlJ1bnRpbWVFeGNlcHRpb26eXwZHCjSD5QIAAHhyABNqYXZhLmxhbmcuRXhjZXB0aW9u0P0fPho7HMQCAAB4cgATamF2YS5sYW5nLlRocm93YWJsZdXGNSc5d7jLAwAETAAFY2F1c2V0ABVMamF2YS9sYW5nL1Rocm93YWJsZTtMAA1kZXRhaWxNZXNzYWdldAASTGphdmEvbGFuZy9TdHJpbmc7WwAKc3RhY2tUcmFjZXQAHltMamF2YS9sYW5nL1N0YWNrVHJhY2VFbGVtZW50O0wAFHN1cHByZXNzZWRFeGNlcHRpb25zdAAQTGphdmEvdXRpbC9MaXN0O3hwc3IANGNvbS5jb21wbGV4aWJsZS5zdGFyZG9nLnBsYW4uZXZhbC5FeGVjdXRpb25FeGNlcHRpb26PoLVSlEBzZQIAAHhyADxjb20uY29tcGxleGlibGUuc3RhcmRvZy5wbGFuLmV2YWwub3BlcmF0b3IuT3BlcmF0b3JFeGNlcHRpb25jJsNYPw/DgQIAAVoAB21Mb2dnZWR4cgAoY29tLmNvbXBsZXhpYmxlLnN0YXJkb2cuU3RhcmRvZ0V4Y2VwdGlvbhti3Dde8jONAgAAeHEAfgABc3IAN2NvbS5jb21wbGV4aWJsZS5zdGFyZG9nLnBsYW4ucGFyc2VyLlF1ZXJ5UGFyc2VFeGNlcHRpb27MeVMSWqwrCAIAAHhyACZjb20uc3RhcmRvZy5zdGFyay5xdWVyeS5NYWxmb3JtZWRRdWVyeckxpgF/S1BEAgAAeHEAfgABc3IAQWNvbS5jb21wbGV4aWJsZS5jb21tb24ucmRmLnF1ZXJ5LnBhcnNlci5zcGFycWwuYXN0LlBhcnNlRXhjZXB0aW9uAAAAAAAAAAECAANMAAxjdXJyZW50VG9rZW50ADpMY29tL2NvbXBsZXhpYmxlL2NvbW1vbi9yZGYvcXVlcnkvcGFyc2VyL3NwYXJxbC9hc3QvVG9rZW47WwAWZXhwZWN0ZWRUb2tlblNlcXVlbmNlc3QAA1tbSVsACnRva2VuSW1hZ2V0ABNbTGphdmEvbGFuZy9TdHJpbmc7eHEAfgACcQB+ABR0AjJFbmNvdW50ZXJlZCAiICIsIiAiLCAiIiBhdCBsaW5lIDE5MCwgY29sdW1uIDMyLgpXYXMgZXhwZWN0aW5nIG9uZSBvZjoKICAgICJ9IiAuLi4KICAgICI8PCIgLi4uCiAgICAiXl4iIC4uLgogICAgInRydWUiIC4uLgogICAgImZhbHNlIiAuLi4KICAgICJVTkRFRiIgLi4uCiAgICA8UV9JUklfUkVGPiAuLi4KICAgIDxQTkFNRV9OUz4gLi4uCiAgICA8UE5BTUVfTE4+IC4uLgogICAgPExBTkdUQUc+IC4uLgogICAgPElOVEVHRVI+IC4uLgogICAgPElOVEVHRVJfUE9TSVRJVkU+IC4uLgogICAgPElOVEVHRVJfTkVHQVRJVkU+IC4uLgogICAgPERFQ0lNQUw+IC4uLgogICAgPERFQ0lNQUxfUE9TSVRJVkU+IC4uLgogICAgPERFQ0lNQUxfTkVHQVRJVkU+IC4uLgogICAgPERPVUJMRT4gLi4uCiAgICA8RE9VQkxFX1BPU0lUSVZFPiAuLi4KICAgIDxET1VCTEVfTkVHQVRJVkU+IC4uLgogICAgPFNUUklOR19MSVRFUkFMMT4gLi4uCiAgICA8U1RSSU5HX0xJVEVSQUwyPiAuLi4KICAgIDxTVFJJTkdfTElURVJBTF9MT05HMT4gLi4uCiAgICA8U1RSSU5HX0xJVEVSQUxfTE9ORzI+IC4uLgogICAgdXIAHltMamF2YS5sYW5nLlN0YWNrVHJhY2VFbGVtZW50OwJGKjw8/SI5AgAAeHAAAABIc3IAG2phdmEubGFuZy5TdGFja1RyYWNlRWxlbWVudGEJxZomNt2FAgAIQgAGZm9ybWF0SQAKbGluZU51bWJlckwAD2NsYXNzTG9hZGVyTmFtZXEAfgAFTAAOZGVjbGFyaW5nQ2xhc3NxAH4ABUwACGZpbGVOYW1lcQB+AAVMAAptZXRob2ROYW1lcQB+AAVMAAptb2R1bGVOYW1lcQB+AAVMAA1tb2R1bGVWZXJzaW9ucQB+AAV4cAEAACv4dAADYXBwdABEY29tLmNvbXBsZXhpYmxlLmNvbW1vbi5yZGYucXVlcnkucGFyc2VyLnNwYXJxbC5hc3QuU3ludGF4VHJlZUJ1aWxkZXJ0ABZTeW50YXhUcmVlQnVpbGRlci5qYXZhdAAWZ2VuZXJhdGVQYXJzZUV4Y2VwdGlvbnBwc3EAfgAYAQAAK2VxAH4AGnEAfgAbcQB+ABx0ABBqal9jb25zdW1lX3Rva2VucHBzcQB+ABgBAAAHiXEAfgAacQB+ABtxAH4AHHQAEElubGluZURhdGFPbmVWYXJwcHNxAH4AGAEAAAdXcQB+ABpxAH4AG3EAfgAcdAAJRGF0YUJsb2NrcHBzcQB+ABgBAAAHPXEAfgAacQB+ABtxAH4AHHQACklubGluZURhdGFwcHNxAH4AGAEAAAzCcQB+ABpxAH4AG3EAfgAcdAAPU2NvcGVUZXJtaW5hdG9ycHBzcQB+ABgBAAALJXEAfgAacQB+ABtxAH4AHHQABlNjb3Blc3Bwc3EAfgAYAQAAC15xAH4AGnEAfgAbcQB+ABxxAH4AKXBwc3EAfgAYAQAAC15xAH4AGnEAfgAbcQB+ABxxAH4AKXBwc3EAfgAYAQAAC15xAH4AGnEAfgAbcQB+ABxxAH4AKXBwc3EAfgAYAQAAC15xAH4AGnEAfgAbcQB+ABxxAH4AKXBwc3EAfgAYAQAAC7NxAH4AGnEAfgAbcQB+ABxxAH4AKXBwc3EAfgAYAQAACvVxAH4AGnEAfgAbcQB+ABx0ABFHcm91cEdyYXBoUGF0dGVybnBwc3EAfgAYAQAADRdxAH4AGnEAfgAbcQB+ABx0ABhHcm91cE9yVW5pb25HcmFwaFBhdHRlcm5wcHNxAH4AGAEAAAyycQB+ABpxAH4AG3EAfgAccQB+ACdwcHNxAH4AGAEAAAuMcQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAuzcQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAtecQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAuzcQB+ABpxAH4AG3EAfgAccQB+AClwcHNxAH4AGAEAAAr1cQB+ABpxAH4AG3EAfgAccQB+ADBwcHNxAH4AGAEAAAW3cQB+ABpxAH4AG3EAfgAcdAALV2hlcmVDbGF1c2VwcHNxAH4AGAEAAAMMcQB+ABpxAH4AG3EAfgAcdAALU2VsZWN0UXVlcnlwcHNxAH4AGAEAAAG9cQB+ABpxAH4AG3EAfgAcdAAFUXVlcnlwcHNxAH4AGAEAAADZcQB+ABpxAH4AG3EAfgAcdAAOUXVlcnlDb250YWluZXJwcHNxAH4AGAEAAAAfcQB+ABpxAH4AG3EAfgAcdAAKcGFyc2VRdWVyeXBwc3EAfgAYAQAAAi5xAH4AGnQAM2NvbS5jb21wbGV4aWJsZS5zdGFyZG9nLnBsYW4ucGFyc2VyLlF1ZXJ5UGFyc2VySW1wbHQAFFF1ZXJ5UGFyc2VySW1wbC5qYXZhcQB+AFFwcHNxAH4AGAEAAAIpcQB+ABpxAH4AU3EAfgBUcQB+AFFwcHNxAH4AGAEAAACJcQB+ABp0AChjb20uY29tcGxleGlibGUuc3RhcmRvZy5wbGFuLlBsYW5Db250ZXh0dAAQUGxhbkNvbnRleHQuamF2YXQAB3BsYW5Gb3JwcHNxAH4AGAEAAAJWcQB+ABp0AC1jb20uY29tcGxleGlibGUuc3RhcmRvZy5wbGFuLmV2YWwuUXVlcnlFbmdpbmV0ABBRdWVyeUVuZ2luZS5qYXZhcQB+AFlwcHNxAH4AGAEAAAGFcQB+ABpxAH4AW3EAfgBcdAAXbGFtYmRhJGV4ZWN1dGVTZWxlY3QkMTZwcHNxAH4AGAEAAALScQB+ABpxAH4AW3EAfgBcdAARZ2V0RXhlY3V0YWJsZVBsYW5wcHNxAH4AGAEAAAHdcQB+ABpxAH4AW3EAfgBcdAALZXhlY3V0ZVBsYW5wcHNxAH4AGAEAAADtcQB+ABpxAH4AW3EAfgBccQB+AGJwcHNxAH4AGAEAAAHDcQB+ABpxAH4AW3EAfgBccQB+AGJwcHNxAH4AGAEAAAGWcQB+ABpxAH4AW3EAfgBcdAARZXhlY3V0ZVNlbGVjdFBsYW5wcHNxAH4AGAEAAAGFcQB+ABpxAH4AW3EAfgBcdAANZXhlY3V0ZVNlbGVjdHBwc3EAfgAYAQAAAQ5xAH4AGnQAPWNvbS5jb21wbGV4aWJsZS5zdGFyZG9nLnF1ZXJ5LkRlZmF1bHRRdWVyeUZhY3RvcnkkU2VsZWN0UXVlcnl0ABhEZWZhdWx0UXVlcnlGYWN0b3J5LmphdmF0AAdleGVjdXRlcHBzcQB+ABgBAAABAHEAfgAacQB+AGpxAH4Aa3EAfgBscHBzcQB+ABgBAAAARHEAfgAadAAtY29tLmNvbXBsZXhpYmxlLnN0YXJkb2cucXVlcnkuRGVsZWdhdGluZ1F1ZXJ5dAAURGVsZWdhdGluZ1F1ZXJ5LmphdmFxAH4AbHBwc3EAfgAYAQAABuZxAH4AGnQAPWNvbS5jb21wbGV4aWJsZS5zdGFyZG9nLmRiLkRhdGFiYXNlSW1wbCRUaW1lclBhdXNpbmdSZWFkUXVlcnl0ABFEYXRhYmFzZUltcGwuamF2YXEAfgBscHBzcQB+ABgBAAAG2XEAfgAacQB+AHJxAH4Ac3EAfgBscHBzcQB+ABgBAAAAPHEAfgAacQB+AG9xAH4AcHEAfgBscHBzcQB+ABgBAAAS9HEAfgAadAA2Y29tLmNvbXBsZXhpYmxlLnN0YXJkb2cuU3RhcmRvZ0tlcm5lbCRTZWN1cmVkUmVhZFF1ZXJ5dAASU3RhcmRvZ0tlcm5lbC5qYXZhcQB+AGxwcHNxAH4AGAEAABLncQB+ABpxAH4Ad3EAfgB4cQB+AGxwcHNxAH4AGAEAAAA8cQB+ABpxAH4Ab3EAfgBwcQB+AGxwcHNxAH4AGAEAAAA2cQB+ABp0ADljb20uY29tcGxleGlibGUuc3RhcmRvZy5Nb25pdG9yZWRRdWVyeSRNb25pdG9yZWRSZWFkUXVlcnl0ABNNb25pdG9yZWRRdWVyeS5qYXZhcQB+AGxwcHNxAH4AGAEAAAAkcQB+ABpxAH4AfHEAfgB9cQB+AGxwcHNxAH4AGAEAAAIgcQB+ABp0ADtjb20uY29tcGxleGlibGUuc3RhcmRvZy5wcm90b2NvbHMuaHR0cC5zZXJ2ZXIuUHJvdG9jb2xVdGlsc3QAElByb3RvY29sVXRpbHMuamF2YXQAEGV4ZWN1dGVSZWFkUXVlcnlwcHNxAH4AGAEAAAIFcQB+ABpxAH4AgHEAfgCBcQB+AIJwcHNxAH4AGAEAAAD6cQB+ABp0ADxjb20uY29tcGxleGlibGUuc3RhcmRvZy5wcm90b2NvbHMuaHR0cC5zZXJ2ZXIuU1BBUlFMUHJvdG9jb2x0ABNTUEFSUUxQcm90b2NvbC5qYXZhdAAMZXhlY3V0ZVF1ZXJ5cHBzcQB+ABgBAAAAi3EAfgAacQB+AIVxAH4AhnQADmdldFNwYXJxbFF1ZXJ5cHBzcQB+ABgBAAAAwHEAfgAadAA0Y29tLnN0YXJkb2cuaHR0cC5zZXJ2ZXIudW5kZXJ0b3cuamF4cnMuRXh0cmFjdFJvdXRlc3QAEkV4dHJhY3RSb3V0ZXMuamF2YXQAEWxhbWJkYSRoYW5kbGVJdCQ1cHBzcQB+ABgBAAAAeHEAfgAadAAwb3JnLmFwYWNoZS5zaGlyby5zdWJqZWN0LnN1cHBvcnQuU3ViamVjdFJ1bm5hYmxldAAUU3ViamVjdFJ1bm5hYmxlLmphdmF0AAVkb1J1bnBwc3EAfgAYAQAAAGxxAH4AGnEAfgCPcQB+AJB0AANydW5wcHNxAH4AGAIAAARocHQAJ2phdmEudXRpbC5jb25jdXJyZW50LlRocmVhZFBvb2xFeGVjdXRvcnQAF1RocmVhZFBvb2xFeGVjdXRvci5qYXZhdAAJcnVuV29ya2VydAAJamF2YS5iYXNldAAHMTEuMC4xOXNxAH4AGAIAAAJ0cHQALmphdmEudXRpbC5jb25jdXJyZW50LlRocmVhZFBvb2xFeGVjdXRvciRXb3JrZXJxAH4AlnEAfgCTcQB+AJhxAH4AmXNxAH4AGAIAAAM9cHQAEGphdmEubGFuZy5UaHJlYWR0AAtUaHJlYWQuamF2YXEAfgCTcQB+AJhxAH4AmXNyAB9qYXZhLnV0aWwuQ29sbGVjdGlvbnMkRW1wdHlMaXN0ergXtDynnt4CAAB4cHhzcgA4Y29tLmNvbXBsZXhpYmxlLmNvbW1vbi5yZGYucXVlcnkucGFyc2VyLnNwYXJxbC5hc3QuVG9rZW4AAAAAAAAAAQIACEkAC2JlZ2luQ29sdW1uSQAJYmVnaW5MaW5lSQAJZW5kQ29sdW1uSQAHZW5kTGluZUkABGtpbmRMAAVpbWFnZXEAfgAFTAAEbmV4dHEAfgARTAAMc3BlY2lhbFRva2VucQB+ABF4cAAAABsAAAC+AAAAHwAAAL4AAADEdAAFImZvbyJzcQB+AKEAAAAgAAAAvgAAACAAAAC+AAAADHQAASxwcHB1cgADW1tJF/fkTxmPiTwCAAB4cAAAABd1cgACW0lNumAmduqypQIAAHhwAAAAAQAAAAZ1cQB+AKgAAAABAAAACXVxAH4AqAAAAAEAAAAedXEAfgCoAAAAAQAAAD91cQB+AKgAAAABAAAATXVxAH4AqAAAAAEAAABodXEAfgCoAAAAAQAAAKx1cQB+AKgAAAABAAAArnVxAH4AqAAAAAEAAACvdXEAfgCoAAAAAQAAALN1cQB+AKgAAAABAAAAtHVxAH4AqAAAAAEAAAC1dXEAfgCoAAAAAQAAALZ1cQB+AKgAAAABAAAAt3VxAH4AqAAAAAEAAAC6dXEAfgCoAAAAAQAAALt1cQB+AKgAAAABAAAAvHVxAH4AqAAAAAEAAADBdXEAfgCoAAAAAQAAAMJ1cQB+AKgAAAABAAAAw3VxAH4AqAAAAAEAAADEdXEAfgCoAAAAAQAAAMV1cQB+AKgAAAABAAAAxnVyABNbTGphdmEubGFuZy5TdHJpbmc7rdJW5+kde0cCAAB4cAAAANp0AAU8RU9GPnQACTxXU19DSEFSPnQADDxXSElURVNQQUNFPnQAAyIoInQAAyIpInQAAyJ7InQAAyJ9InQAAyJbInQAAyJdInQABCI8PCJ0AAQiPj4idAADIjsidAADIiwidAADIi4idAADIj0idAAEIiE9InQAAyI+InQAAyI8InQABCI8PSJ0AAQiPj0idAADIiEidAAEInx8InQABCImJiJ0AAMiKyJ0AAMiLSJ0AAMiKiJ0AAMiPyJ0AAMiLyJ0AAMifCJ0AAMiXiJ0AAQiXl4idAAFPE5JTD50AAY8QU5PTj50AAY8SElOVD50ABU8U0lOR0xFX0xJTkVfQ09NTUVOVD50AAMiYSJ0AAQiYXMidAAEImJ5InQABCJpZiJ0AAQiaW4idAAEInR6InQABSJtZDUidAAFImFnZyJ0AAUiYXNjInQABSJhc2sidAAFInN0ciJ0AAUic3VtInQABSJtaW4idAAFIm1heCJ0AAUiYXZnInQABSJhYnMidAAFImRheSJ0AAUibm93InQABSJJUkkidAAFIlVSSSJ0AAYidGhlbiJ0AAYiY2VpbCJ0AAYieWVhciJ0AAYic2hhMSJ0AAYidXVpZCJ0AAYibGFuZyJ0AAYiYmFzZSJ0AAYiZGVzYyJ0AAYidHJ1ZSJ0AAYiYmluZCJ0AAYiZnJvbSJ0AAYicmFuZCJ0AAYianNvbiJ0AAUic3FsInQACSJncmFwaHFsInQABSJjc3YidAAHIm5hbWVkInQAByJsb2NhbCJ0AAkidmlydHVhbCJ0AAcid2hlcmUidAAHImxpbWl0InQAByJyZWdleCJ0AAciZmFsc2UidAAHIm9yZGVyInQAByJncm91cCJ0AAciZ3JhcGgidAAHInVuaW9uInQAByJib3VuZCJ0AAcidWNhc2UidAAHImxjYXNlInQAByJmbG9vciJ0AAgidW5uZXN0InQADSJhcnJheV9pbmRleCJ0AAciaXNJUkkidAAHImlzVVJJInQAByJwYXRocyJ0AAoic2hvcnRlc3QidAAIImN5Y2xpYyJ0AAgibGVuZ3RoInQAByJzdGFydCJ0AAUiZW5kInQABSJ2aWEidAAHIm1vbnRoInQACCJzYW1wbGUidAAHImhvdXJzInQAByJibm9kZSJ0AAcic3RyZHQidAAHIm1pbnVzInQAByJjb3VudCJ0AAciVU5ERUYidAAIImZpbHRlciJ0AAgiaGF2aW5nInQACCJub3QgaW4idAAIInN0cmxlbiJ0AAgic3Vic3RyInQACCJwcmVmaXgidAAIInNlbGVjdCJ0AAgiY29uY2F0InQACCJzaGEyMjQidAAIInNoYTI1NiJ0AAgic2hhMzg0InQACCJzaGE1MTIidAAIIm9mZnNldCJ0AAgidmFsdWVzInQACCJleGlzdHMidAAJIm1pbnV0ZXMidAAJInNlY29uZHMidAAJInJlZHVjZWQidAAJImlzQmxhbmsidAAJInNlcnZpY2UidAAKImJpbmRpbmdzInQACiJzdHJBZnRlciJ0AAkicmVwbGFjZSJ0AAoiZnVuY3Rpb24idAAKIm9wdGlvbmFsInQACSJzdHJsYW5nInQACSJzdHJ1dWlkInQACSJzdHJFbmRzInQACiJkZXNjcmliZSJ0AAoiZGF0YXR5cGUidAAKImRpc3RpbmN0InQACiJzYW1lVGVybSJ0AAoidGltZXpvbmUidAAKImNvbnRhaW5zInQACiJjb2FsZXNjZSJ0AAwibm90IGV4aXN0cyJ0AAsiY29uc3RydWN0InQACyJzdHJTdGFydHMidAALInN0ckJlZm9yZSJ0AAsiaXNMaXRlcmFsInQACyJpc051bWVyaWMidAALInNlcGFyYXRvciJ0AA0ibGFuZ21hdGNoZXMidAAOImdyb3VwX2NvbmNhdCJ0ABAiZW5jb2RlX2Zvcl9VUkkidAAJIm1hcHBpbmcidAAKInZhbGlkYXRlInQABSJwZXIidAAHInNoYXBlInQACCJzaGFwZXMidAAGImxvYWQidAAHImNsZWFyInQABiJkcm9wInQABSJhZGQidAAGIm1vdmUidAAGImNvcHkidAAIImNyZWF0ZSJ0AAgiaW5zZXJ0InQABiJkYXRhInQACCJkZWxldGUidAAGIndpdGgidAAIInNpbGVudCJ0AAkiZGVmYXVsdCJ0AAUiYWxsInQABiJpbnRvInQABCJ0byJ0AAcidXNpbmcidAALPFFfSVJJX1JFRj50AAo8QUdHX05BTUU+dAAKPFBOQU1FX05TPnQACjxQTkFNRV9MTj50ABI8QkxBTktfTk9ERV9MQUJFTD50AAY8VkFSMT50AAY8VkFSMj50AAk8TEFOR1RBRz50AAk8SU5URUdFUj50ABI8SU5URUdFUl9QT1NJVElWRT50ABI8SU5URUdFUl9ORUdBVElWRT50AAk8REVDSU1BTD50AAo8REVDSU1BTDE+dAAKPERFQ0lNQUwyPnQAEjxERUNJTUFMX1BPU0lUSVZFPnQAEjxERUNJTUFMX05FR0FUSVZFPnQACDxET1VCTEU+dAAJPERPVUJMRTE+dAAJPERPVUJMRTI+dAAJPERPVUJMRTM+dAAKPEVYUE9ORU5UPnQAETxET1VCTEVfUE9TSVRJVkU+dAARPERPVUJMRV9ORUdBVElWRT50ABE8U1RSSU5HX0xJVEVSQUwxPnQAETxTVFJJTkdfTElURVJBTDI+dAAWPFNUUklOR19MSVRFUkFMX0xPTkcxPnQAFjxTVFJJTkdfTElURVJBTF9MT05HMj50AAw8U0FGRV9DSEFSMT50AAw8U0FGRV9DSEFSMj50ABE8U0FGRV9DSEFSX0xPTkcxPnQAETxTQUZFX0NIQVJfTE9ORzI+dAAHPEVDSEFSPnQABTxIRVg+dAAHPEFMUEhBPnQABTxOVU0+dAAPPEZVTkNUSU9OX05BTUU+dAAPPFBOX0NIQVJTX0JBU0U+dAAMPFBOX0NIQVJTX1U+dAAKPFZBUl9DSEFSPnQACjxQTl9DSEFSUz50AAs8UE5fUFJFRklYPnQACjxQTl9MT0NBTD50AAU8UExYPnQACTxQRVJDRU5UPnQADjxQTl9MT0NBTF9FU0M+dAAJPFZBUk5BTUU+dAJISW52YWxpZCBTUEFSUUwgcXVlcnk6IEVuY291bnRlcmVkICIgIiwiICIsICIiIGF0IGxpbmUgMTkwLCBjb2x1bW4gMzIuCldhcyBleHBlY3Rpbmcgb25lIG9mOgogICAgIn0iIC4uLgogICAgIjw8IiAuLi4KICAgICJeXiIgLi4uCiAgICAidHJ1ZSIgLi4uCiAgICAiZmFsc2UiIC4uLgogICAgIlVOREVGIiAuLi4KICAgIDxRX0lSSV9SRUY+IC4uLgogICAgPFBOQU1FX05TPiAuLi4KICAgIDxQTkFNRV9MTj4gLi4uCiAgICA8TEFOR1RBRz4gLi4uCiAgICA8SU5URUdFUj4gLi4uCiAgICA8SU5URUdFUl9QT1NJVElWRT4gLi4uCiAgICA8SU5URUdFUl9ORUdBVElWRT4gLi4uCiAgICA8REVDSU1BTD4gLi4uCiAgICA8REVDSU1BTF9QT1NJVElWRT4gLi4uCiAgICA8REVDSU1BTF9ORUdBVElWRT4gLi4uCiAgICA8RE9VQkxFPiAuLi4KICAgIDxET1VCTEVfUE9TSVRJVkU+IC4uLgogICAgPERPVUJMRV9ORUdBVElWRT4gLi4uCiAgICA8U1RSSU5HX0xJVEVSQUwxPiAuLi4KICAgIDxTVFJJTkdfTElURVJBTDI+IC4uLgogICAgPFNUUklOR19MSVRFUkFMX0xPTkcxPiAuLi4KICAgIDxTVFJJTkdfTElURVJBTF9MT05HMj4gLi4uCiAgICB1cQB+ABYAAAAgc3EAfgAYAQAAAlVxAH4AGnEAfgBTcQB+AFRxAH4AUXBwc3EAfgAYAQAAAilxAH4AGnEAfgBTcQB+AFRxAH4AUXBwc3EAfgAYAQAAAIlxAH4AGnEAfgBXcQB+AFhxAH4AWXBwc3EAfgAYAQAAAlZxAH4AGnEAfgBbcQB+AFxxAH4AWXBwc3EAfgAYAQAAAYVxAH4AGnEAfgBbcQB+AFxxAH4AXnBwc3EAfgAYAQAAAtJxAH4AGnEAfgBbcQB+AFxxAH4AYHBwc3EAfgAYAQAAAd1xAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAO1xAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAcNxAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAZZxAH4AGnEAfgBbcQB+AFxxAH4AZnBwc3EAfgAYAQAAAYVxAH4AGnEAfgBbcQB+AFxxAH4AaHBwc3EAfgAYAQAAAQ5xAH4AGnEAfgBqcQB+AGtxAH4AbHBwc3EAfgAYAQAAAQBxAH4AGnEAfgBqcQB+AGtxAH4AbHBwc3EAfgAYAQAAAERxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAABuZxAH4AGnEAfgBycQB+AHNxAH4AbHBwc3EAfgAYAQAABtlxAH4AGnEAfgBycQB+AHNxAH4AbHBwc3EAfgAYAQAAADxxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAAEvRxAH4AGnEAfgB3cQB+AHhxAH4AbHBwc3EAfgAYAQAAEudxAH4AGnEAfgB3cQB+AHhxAH4AbHBwc3EAfgAYAQAAADxxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAAADZxAH4AGnEAfgB8cQB+AH1xAH4AbHBwc3EAfgAYAQAAACRxAH4AGnEAfgB8cQB+AH1xAH4AbHBwc3EAfgAYAQAAAiBxAH4AGnEAfgCAcQB+AIFxAH4AgnBwc3EAfgAYAQAAAgVxAH4AGnEAfgCAcQB+AIFxAH4AgnBwc3EAfgAYAQAAAPpxAH4AGnEAfgCFcQB+AIZxAH4Ah3Bwc3EAfgAYAQAAAItxAH4AGnEAfgCFcQB+AIZxAH4AiXBwc3EAfgAYAQAAAMBxAH4AGnEAfgCLcQB+AIxxAH4AjXBwc3EAfgAYAQAAAHhxAH4AGnEAfgCPcQB+AJBxAH4AkXBwc3EAfgAYAQAAAGxxAH4AGnEAfgCPcQB+AJBxAH4Ak3Bwc3EAfgAYAgAABGhwcQB+AJVxAH4AlnEAfgCXcQB+AJhxAH4AmXNxAH4AGAIAAAJ0cHEAfgCbcQB+AJZxAH4Ak3EAfgCYcQB+AJlzcQB+ABgCAAADPXBxAH4AnXEAfgCecQB+AJNxAH4AmHEAfgCZcQB+AKB4cQB+AZx1cQB+ABYAAAAbc3EAfgAYAQAAATRxAH4AGnEAfgBbcQB+AFx0AAtoYW5kbGVFcnJvcnBwc3EAfgAYAQAAAedxAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAO1xAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAcNxAH4AGnEAfgBbcQB+AFxxAH4AYnBwc3EAfgAYAQAAAZZxAH4AGnEAfgBbcQB+AFxxAH4AZnBwc3EAfgAYAQAAAYVxAH4AGnEAfgBbcQB+AFxxAH4AaHBwc3EAfgAYAQAAAQ5xAH4AGnEAfgBqcQB+AGtxAH4AbHBwc3EAfgAYAQAAAQBxAH4AGnEAfgBqcQB+AGtxAH4AbHBwc3EAfgAYAQAAAERxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAABuZxAH4AGnEAfgBycQB+AHNxAH4AbHBwc3EAfgAYAQAABtlxAH4AGnEAfgBycQB+AHNxAH4AbHBwc3EAfgAYAQAAADxxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAAEvRxAH4AGnEAfgB3cQB+AHhxAH4AbHBwc3EAfgAYAQAAEudxAH4AGnEAfgB3cQB+AHhxAH4AbHBwc3EAfgAYAQAAADxxAH4AGnEAfgBvcQB+AHBxAH4AbHBwc3EAfgAYAQAAADZxAH4AGnEAfgB8cQB+AH1xAH4AbHBwc3EAfgAYAQAAACRxAH4AGnEAfgB8cQB+AH1xAH4AbHBwc3EAfgAYAQAAAiBxAH4AGnEAfgCAcQB+AIFxAH4AgnBwc3EAfgAYAQAAAgVxAH4AGnEAfgCAcQB+AIFxAH4AgnBwc3EAfgAYAQAAAPpxAH4AGnEAfgCFcQB+AIZxAH4Ah3Bwc3EAfgAYAQAAAItxAH4AGnEAfgCFcQB+AIZxAH4AiXBwc3EAfgAYAQAAAMBxAH4AGnEAfgCLcQB+AIxxAH4AjXBwc3EAfgAYAQAAAHhxAH4AGnEAfgCPcQB+AJBxAH4AkXBwc3EAfgAYAQAAAGxxAH4AGnEAfgCPcQB+AJBxAH4Ak3Bwc3EAfgAYAgAABGhwcQB+AJVxAH4AlnEAfgCXcQB+AJhxAH4AmXNxAH4AGAIAAAJ0cHEAfgCbcQB+AJZxAH4Ak3EAfgCYcQB+AJlzcQB+ABgCAAADPXBxAH4AnXEAfgCecQB+AJNxAH4AmHEAfgCZcQB+AKB4AHQCfmNvbS5jb21wbGV4aWJsZS5zdGFyZG9nLnBsYW4uZXZhbC5FeGVjdXRpb25FeGNlcHRpb246IEludmFsaWQgU1BBUlFMIHF1ZXJ5OiBFbmNvdW50ZXJlZCAiICIsIiAiLCAiIiBhdCBsaW5lIDE5MCwgY29sdW1uIDMyLgpXYXMgZXhwZWN0aW5nIG9uZSBvZjoKICAgICJ9IiAuLi4KICAgICI8PCIgLi4uCiAgICAiXl4iIC4uLgogICAgInRydWUiIC4uLgogICAgImZhbHNlIiAuLi4KICAgICJVTkRFRiIgLi4uCiAgICA8UV9JUklfUkVGPiAuLi4KICAgIDxQTkFNRV9OUz4gLi4uCiAgICA8UE5BTUVfTE4+IC4uLgogICAgPExBTkdUQUc+IC4uLgogICAgPElOVEVHRVI+IC4uLgogICAgPElOVEVHRVJfUE9TSVRJVkU+IC4uLgogICAgPElOVEVHRVJfTkVHQVRJVkU+IC4uLgogICAgPERFQ0lNQUw+IC4uLgogICAgPERFQ0lNQUxfUE9TSVRJVkU+IC4uLgogICAgPERFQ0lNQUxfTkVHQVRJVkU+IC4uLgogICAgPERPVUJMRT4gLi4uCiAgICA8RE9VQkxFX1BPU0lUSVZFPiAuLi4KICAgIDxET1VCTEVfTkVHQVRJVkU+IC4uLgogICAgPFNUUklOR19MSVRFUkFMMT4gLi4uCiAgICA8U1RSSU5HX0xJVEVSQUwyPiAuLi4KICAgIDxTVFJJTkdfTElURVJBTF9MT05HMT4gLi4uCiAgICA8U1RSSU5HX0xJVEVSQUxfTE9ORzI+IC4uLgogICAgdXEAfgAWAAAAFnNxAH4AGAEAAAAucQB+ABp0AC1jb20uc3RhcmRvZy5zdGFyay5xdWVyeS5RdWVyeUV4ZWN1dGlvbkZhaWx1cmV0ABpRdWVyeUV4ZWN1dGlvbkZhaWx1cmUuamF2YXQADmZvckxvZ2dlZEVycm9ycHBzcQB+ABgBAAABEXEAfgAacQB+AGpxAH4Aa3EAfgBscHBzcQB+ABgBAAABAHEAfgAacQB+AGpxAH4Aa3EAfgBscHBzcQB+ABgBAAAARHEAfgAacQB+AG9xAH4AcHEAfgBscHBzcQB+ABgBAAAG5nEAfgAacQB+AHJxAH4Ac3EAfgBscHBzcQB+ABgBAAAG2XEAfgAacQB+AHJxAH4Ac3EAfgBscHBzcQB+ABgBAAAAPHEAfgAacQB+AG9xAH4AcHEAfgBscHBzcQB+ABgBAAAS9HEAfgAacQB+AHdxAH4AeHEAfgBscHBzcQB+ABgBAAAS53EAfgAacQB+AHdxAH4AeHEAfgBscHBzcQB+ABgBAAAAPHEAfgAacQB+AG9xAH4AcHEAfgBscHBzcQB+ABgBAAAANnEAfgAacQB+AHxxAH4AfXEAfgBscHBzcQB+ABgBAAAAJHEAfgAacQB+AHxxAH4AfXEAfgBscHBzcQB+ABgBAAACIHEAfgAacQB+AIBxAH4AgXEAfgCCcHBzcQB+ABgBAAACBXEAfgAacQB+AIBxAH4AgXEAfgCCcHBzcQB+ABgBAAAA+nEAfgAacQB+AIVxAH4AhnEAfgCHcHBzcQB+ABgBAAAAi3EAfgAacQB+AIVxAH4AhnEAfgCJcHBzcQB+ABgBAAAAwHEAfgAacQB+AItxAH4AjHEAfgCNcHBzcQB+ABgBAAAAeHEAfgAacQB+AI9xAH4AkHEAfgCRcHBzcQB+ABgBAAAAbHEAfgAacQB+AI9xAH4AkHEAfgCTcHBzcQB+ABgCAAAEaHBxAH4AlXEAfgCWcQB+AJdxAH4AmHEAfgCZc3EAfgAYAgAAAnRwcQB+AJtxAH4AlnEAfgCTcQB+AJhxAH4AmXNxAH4AGAIAAAM9cHEAfgCdcQB+AJ5xAH4Ak3EAfgCYcQB+AJlxAH4AoHgB"}
`

nb: would it be easy to include failing SPARQL and/or GQL queries in the debug panel?

bprusinowski commented:
Oups, I was too fast 😄 Should be fine now!

bprusinowski commented:
@Rdataflow I would need to investigate this, I think it would take some time to also show failing queries in the debug panel, as now if the query fails, it fails inside the decorated SPARQL client which does not add the query to the list of queries then; I don't know right away how to handle this so they are also showed. This should be possible but probably would be a bit bigger thing 👀

Rdataflow commented:
> @Rdataflow I would need to investigate this, I think it would take some time to also show failing queries in the debug panel, as now if the query fails, it fails inside the decorated SPARQL client which does not add the query to the list of queries then; I don't know right away how to handle this so they are also showed. This should be possible but probably would be a bit bigger thing 👀

fine, it's not a prio for now

Rdataflow commented:
@bprusinowski well done for the multiple keywords 👍 
- [ ] WRT #979 a user expects to see every single non-expired version, thus no need to ask for `versionHistory` and group within the resolver, i.e. [here_we_expect_2_results--one_published_and_one_draft](https://visualization-tool-git-perf-optimize-rdf-queries-ixt1.vercel.app/browse/organization/https%3A%2F%2Fregister.ld.admin.ch%2Fopendataswiss%2Forg%2Fbundesamt-fur-energie-bfe?includeDrafts=true&order=SCORE&search=zero&dataSource=Prod) (in sync with @ortnever)
- [ ] better query perf on keywords: see this pattern https://s.zazuko.com/2vF5RGX to receive results 😄

bprusinowski commented:
@Rdataflow it looks like we still need to keep grouping by version history, as otherwise the number of datasets on PROD increases from 58 to 120+ (without including draft datasets). We need to use the most recent version of the cube to only display one entry for given iri; but to fix the error you mentioned, we just need to add publication status to the grouping logic and then we'll show both published and draft versions of the same cube – I've added a commit to fix this.

bprusinowski commented:
Thanks for the example query @Rdataflow, now it works perfectly :)

Rdataflow commented:
@bprusinowski well done for the fix on multiple keywords. 
- [ ] filter 

> @Rdataflow it looks like we still need to keep grouping by version history, as otherwise the number of datasets on PROD increases from 58 to 120+ (without including draft datasets). We need to use the most recent version of the cube to only display one entry for a given iri; but to fix the error you mentioned, we just need to add publication status to the grouping logic and then we'll show both published and draft versions of the same cube – I've added a commit to fix this.

Of course you still get dups for each cubeIri (dup theme etc.) https://s.zazuko.com/LyiXEv
So even w/o history you still need to regroup by cubeIri ... https://s.zazuko.com/LyiXEv

bprusinowski commented:
Yes 👍 But still, isn't it safer to use version history here? Or is there no difference in this case at all?

bprusinowski commented:
@Rdataflow after checking the code it looks like the support of multiple creators needs some thinking before it's implemented on our side, I've opened [an issue](https://github.com/visualize-admin/visualization-tool/issues/1238) to keep track of it.

I think the remaining points we discussed should be done now 👍

Rdataflow commented:
@bprusinowski searching for `foobar` throws an error...  https://visualization-tool-git-perf-optimize-rdf-queries-ixt1.vercel.app/browse?includeDrafts=false&order=SCORE&search=foo&dataSource=Prod

BTW: wanna include a test for this case?

ptbrowne commented on app/browser/dataset-browse.tsx:
```diff
@@ -622,29 +621,23 @@ export const SearchFilters = ({ data }: { data?: DataCubesQuery }) => {
+      const countables = [...dataCube.themes, dataCube.creator].filter(truthy);
+
+      for (const { iri } of countables) {
+        result[iri] = (result[iri] ?? 0) + 1;
```
I think the idea behind the count was to not have to retrieve all the matching data cubes in case there was a really long list of data cubes. Isn't it a problem now ?

ptbrowne replied:
What's weird to me is that counting if done via SPARLQ request should not be done via another query from the front-end, and should be done on the server as part of the search query. Don't know why I did it this way before 🤷 .

bprusinowski replied:
I think that could become a problem if we implement pagination, and put a limit on the number of cubes we fetch at once; then we would need to have a way to retrieve the total counts separately 👍 

However now that the search query takes ~1.5s to fetch 263 cubes (INT + drafts), and we only have 64 on PROD, we could even do the pagination purely on the client side, and move to server side once it becomes a performance problem. But we didn't discuss pagination at any point yet, and in the current form this query was kinda an "extra one" 😄

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -183,9 +223,8 @@ export const searchCubes = async ({
-
-          )`
+        .join(" || ")})
+      )`
```
Cool to have only 1 query ! Do you remember the time savings ? Don't know why I separated it into to again 😓

bprusinowski replied:
The removed DESCRIBE query was taking around 200ms, and the overall improvement is hard to measure yet, but e.g. currently on PROD the search query without any filters takes 1.5s, and in the preview deployment it's 0.15s 👀 But it also might depend on the deployment itself (locations of servers, etc), so I guess we'll fully see when it's deployed to TEST :)

ptbrowne commented on app/domain/data.ts:
```diff
@@ -70,6 +70,17 @@ export type GeoData = {
   symbolLayer: SymbolLayer | undefined;
 };
 
+// Extracted for performance reasons.
```
Curiosity : What was the problem ? Where was it extracted from ?

bprusinowski replied:
As here we were fighting for every ms, I wanted to avoid the resolver chain with additional queries, as basically our SearchCube was needed in its full form, so we didn't need to have a way to e.g. only specify themes or labels (see the sparqlEditorUrl, data, sparqlEditorUrl).

<img width="307" alt="Screenshot 2023-10-31 at 09 12 29" src="https://github.com/visualize-admin/visualization-tool/assets/52032047/327d6084-9bb5-4112-9785-815e34324062">

Maybe we didn't get a lot of performance improvement here, but with this + making it a Scalar, we only send one query.

<img width="137" alt="Screenshot 2023-10-31 at 09 15 05" src="https://github.com/visualize-admin/visualization-tool/assets/52032047/2c6644ff-28cc-4540-8e84-e69dbc695014">

ptbrowne replied:
Understood, thanks !

ptbrowne commented on app/graphql/schema.graphql:
```diff
@@ -323,18 +323,7 @@ type OrdinalMeasure implements Dimension {
-  datePublished: String
-  themes: SearchCubeThemes!
-}
+scalar SearchCube
```
I don't quite understand why it's good to use a scalar vs a type here ? What is the difference ?

bprusinowski replied:
Scalar was used to stop resolvers chain, as we already have all the data we need at this level, and could avoid additional queries by telling Apollo we want to stop right here :)

ptbrowne replied:
Yes perfect, I did not have the "resolver chain stopping" in my mind. Super nice perf improvement.

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -34,36 +31,6 @@ const makeInFilter = (varName: string, values: string[]) => {
-    data,
-  };
-};
-
```
What is defined somewhere else ?

bprusinowski replied:
I think this might have been implemented before SPARQL clients were decorated? I was surprised that we send two SearchCubes queries every time, and I discovered that in fact we pushed the same query using this logic to the query we already saved using the timed SPARQL client 👀

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -151,13 +153,26 @@ export const searchCubes = async ({
+          "subthemeLabel",
+          { locale }
+        )}
+      }
```
Why moving it here ?

bprusinowski replied:
It was moved here to avoid a need for an additional query; basically to fetch everything related to cube metadata directly inside this one query

ptbrowne commented on app/browser/dataset-browse.tsx:
```diff
@@ -353,12 +353,12 @@ const NavItem = ({
-    );
+    return `/browse/${newFilters
+      .map(encodeFilter)
+      .join("/")}?${extraURLParams}`;
```
Why no more theme & organisation filter ?

bprusinowski replied:
If I remember correctly, this was needed to avoid the below behavior and to make it possible to only de-select one filter and make the other one stay :)


https://github.com/visualize-admin/visualization-tool/assets/52032047/2d65079d-8f6d-43ad-8182-9ed5f4f4a5f3

bprusinowski replied:
But the filters still work :D

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -147,12 +143,21 @@ export const searchCubes = async ({
+          ${buildLocalizedSubQuery("themeIri", "schema:name", "themeLabel", {
+            locale,
+          })}
+        }
```
Curiosity : Why do we need another graph for themes ?

bprusinowski replied:
I think it's better for the performance if we narrow down the graph, so it doesn't need to look everywhere to find the things? @Rdataflow might have a better understanding here :)

Rdataflow replied:
@bprusinowski well understood 😄 it's about narrowing down to one `GRAPH` (a.k.a. one partition of the SPARQL database)

ptbrowne commented on app/rdf/query-search-score-utils.ts:
```diff
@@ -30,6 +30,7 @@ export const computeScores = (
 
   if (query) {
     for (const cube of cubes) {
+      // If a cube have been found, it has at least a score of 1.
```
`has been` :)

ptbrowne commented on e2e/search.spec.ts:
```diff
@@ -121,3 +122,88 @@ test("sort order", async ({ page, selectors, screen, actions }) => {
+  expect(count5).toEqual(count6);
+  expect(count5).toEqual(count7);
+  expect(count5).toEqual(count8);
+});
```
Nice to have added more tests !

ptbrowne commented on app/browser/dataset-browse.tsx:
```diff
@@ -596,20 +596,22 @@ const NavSection = ({
       sourceUrl: dataSource.url,
       locale,
     },
+    pause: !!dataset,
```
👍

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -222,28 +218,17 @@ export const searchCubes = async ({
+          CONTAINS(LCASE(?_themeLabel), LCASE(?keyword)) ||
+          CONTAINS(LCASE(?_subthemeLabel), LCASE(?keyword)) ||
+          CONTAINS(LCASE(?_publisher), LCASE(?keyword))
+        )`
```
❤️

ptbrowne commented on app/rdf/query-search.ts:
```diff
@@ -195,44 +195,50 @@ export const searchCubes = async ({
+        CONTAINS(LCASE(?subthemeLabel_),   LCASE(?keyword)) || 
+          
+        CONTAINS(LCASE(?publisher), LCASE(?keyword))
+      )`
```
Beautiful query now !!

ptbrowne reviewed (approved):
👏 👏 LGTM, lots of small questions since I am not deep inside the SPARQL, sometimes the changes are a bit obscure to me.

Rdataflow commented on app/rdf/query-search.ts:
```diff
@@ -238,11 +237,16 @@ export const searchCubes = async ({
     .BY`?creatorLabel`.prologue`${pragmas}`;
 
+  if (themeValues.length) {
+    scoresQuery = scoresQuery.HAVING`${themeValues
```
@bprusinowski what was the issue with `FILTER()` ? - can we find a fix for that one?

bprusinowski replied:
But it doesn't look like it affected the performance in any noticeable way

Rdataflow replied:
OK, you don't want to loose the other themes on the selected cubes, right?
Instead of bad performing `HAVING` you may state the theme i.e. `?iri dcat:theme <https://register.ld.admin.ch/opendataswiss/category/agriculture> .`
this will bring the desired filtering effect, not?

bprusinowski replied:
Yup, you are right – I updated the query 🎉

bprusinowski commented:
@Rdataflow I'm going to merge the PR to main, as it's getting quite big and it looks like we completed every task and now have only one fast query to fetch the cubes 🎉 

I would propose to then test on TEST and implement any potential fixes / improvements in smaller PRs. What do you think?

Rdataflow commented:
@bprusinowski well done 🎉  let's go and test on TEST

Rdataflow commented:
@bprusinowski 
- [ ] fix needed for error   see https://test.visualize.admin.ch/browse?includeDrafts=true&dataSource=Prod
    ```
    [GraphQL] Unknown type "DataCubeResultOrder". Did you mean "SearchCubeResultOrder"?
    [GraphQL] Unknown type "DataCubeSearchFilter".
    [GraphQL] Cannot query field "dataCubes" on type "Query".

bprusinowski commented:
@Rdataflow the page works fine for me. Did you refresh the page after it deployed?

bprusinowski commented:
The error references old GQL types, before this refactor. It looks like the query sent to new deployment uses these, see https://github.com/visualize-admin/visualization-tool/pull/1236/files#diff-80407bb67b6461b6283843baa5e76d773a5d74e8393c52ac96cdf81ceed1c343

Let me know if the error persists 👀

Rdataflow commented:
so probably a caching issue locally 👍

Rdataflow commented:
FTR - performance of query for CubeList improved by **factor 4** :rocket: 
- p(80) before the PR **~880ms**
   ![image](https://github.com/visualize-admin/visualization-tool/assets/26090254/8970a89c-a2d3-4673-9b92-43fc71202f3c)
- p(80) after the PR **~205ms**
   ![image](https://github.com/visualize-admin/visualization-tool/assets/26090254/1b638d26-d258-45ac-af4b-56e7b9b7db17)

@bprusinowski congrats :tada: 
cc @sosiology @adintegra @ortnever @zellersabine @claudiodigallo

bprusinowski commented:
Fantastic news, thanks again @Rdataflow for your support with optimizations 💯
